python/dgl/distributed/sparse_emb/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,208 @@
+        the values of the embeddings are initialized to zero.
+    part_policy : PartitionPolicy, optional
+        The partition policy that assigns embeddings to different machines in the cluster.
+        Currently, it only supports node partition policy or edge partition policy.
+        The system determines the right partition policy automatically.
+
+    Examples
+    --------
+    >>> def initializer(shape, dtype):
+            arr = th.zeros(shape, dtype=dtype)
+            arr.uniform_(-1, 1)
+            return arr
+    >>> emb = dgl.distributed.NodeEmbedding(g.number_of_nodes(), 10, init_func=initializer)
+    >>> optimizer = dgl.distributed.optim.SparseAdagrad([emb], lr=0.001)
+    >>> for blocks in dataloader:
+    ...     feats = emb(nids)
+    ...     loss = F.sum(feats + 1, 0)
+    ...     loss.backward()
+    ...     optimizer.step()
+
+    Note
+    ----
+    When a ``NodeEmbedding``  object is used when the deep learning framework is recording
+    the forward computation, users have to invoke
+    py:meth:`~dgl.distributed.optim.SparseAdagrad.step` afterwards. Otherwise, there will be
+    some memory leak.
+    '''
+    def __init__(self, num_embeddings, embedding_dim, name=None,
+                 init_func=None, part_policy=None):
+        self._tensor = DistTensor((num_embeddings, embedding_dim), F.float32, name,
+                                  init_func=init_func, part_policy=part_policy)
+        self._trace = []
+        self._name = name
+        self._num_embeddings = num_embeddings
+        self._embedding_dim = embedding_dim
+
+        # Check whether it is multi-gpu/distributed training or not
+        if th.distributed.is_initialized():
+            self._rank = th.distributed.get_rank()
+            self._world_size = th.distributed.get_world_size()
+        else:
+            assert 'th.distributed shoud be initialized'
+        self._optm_state = None # track optimizer state
+        self._part_policy = part_policy
+
+    def __call__(self, idx, device=th.device('cpu')):
+        """
+        node_ids : th.tensor
+            Index of the embeddings to collect.
+        device : th.device
+            Target device to put the collected embeddings.
+
+        Returns
+        -------
+        Tensor
+            The requested node embeddings
+        """
+        idx = utils.toindex(idx).tousertensor()
+        emb = self._tensor[idx].to(device, non_blocking=True)
+        if F.is_recording():
+            emb = F.attach_grad(emb)
+            self._trace.append((idx.to(device, non_blocking=True), emb))
+        return emb
+
+    def reset_trace(self):
+        '''Reset the traced data.
+        '''
+        self._trace = []
+
+    @property
+    def part_policy(self):
+        """Return the partition policy
+
+        Returns
+        -------
+        PartitionPolicy
+            partition policy
+        """
+        return self._part_policy
+
+    @property
+    def rank(self):
```

Comment:
i don't understand why we should keep this field in Embedding? is it used in any place? I think it'll confuse users because this rank is different from the one in DistGraph.

Reply:
They are not used. I will delete them.